src/CoreWCF.Primitives/tests/ServiceKnownTypeTests.cs

```diff
@@ -0,0 +1,223 @@
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            Assert.Throws<System.Xml.XmlException>(() =>
+            {
+                BaseMsg msg = channel.Ping("hello");
+            });
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest1()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService1, PingService1>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest2()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService2, PingService2>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest3()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService3, PingService3>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest4()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService4, PingService4>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+    }
+
+    internal class PingServiceEx : IPingServiceEx
+    {
+        public BaseMsg Ping(string msg)
+        {
+            return new DerivedMsg { msg = msg };
+        }
+
+    }
+
+    [ServiceContract]
+    [System.ServiceModel.ServiceContract]
+    public interface IPingServiceEx
+    {
+        [OperationContract]
+        [System.ServiceModel.OperationContract]
+        BaseMsg Ping(string msg);
+
+    }
+
+    internal class PingService1 : IPingService1
+    {
+        public BaseMsg Ping(string msg)
+        {
+            return new DerivedMsg { msg = msg };
+        }
+
+    }
+
+    [ServiceContract]
+    [System.ServiceModel.ServiceContract]
+    [ServiceKnownType(typeof(DerivedMsg))]
+    [System.ServiceModel.ServiceKnownType(typeof(DerivedMsg))]
+    public interface IPingService1
+    {
+        [OperationContract]
+        [System.ServiceModel.OperationContract]
+        BaseMsg Ping(string msg);
+
+    }
```

Comment:
There are extra blank lines before the close of the interface and the implementing classes (which should be refactored to one liners anyway).

Reply:
Should look Ok after refactoring